tests/src/test/java/io/streamnative/pulsar/handlers/kop/MessagePublishThrottlingTest.java

```diff
@@ -0,0 +1,285 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.streamnative.pulsar.handlers.kop;
+
+
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotEquals;
+import static org.testng.Assert.assertTrue;
+
+import com.google.common.collect.Sets;
+import java.util.Properties;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import lombok.Cleanup;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.serialization.ByteArraySerializer;
+import org.apache.kafka.common.serialization.IntegerSerializer;
+import org.apache.pulsar.broker.service.Producer;
+import org.apache.pulsar.broker.service.PublishRateLimiter;
+import org.apache.pulsar.broker.service.persistent.PersistentTopic;
+import org.apache.pulsar.common.naming.TopicName;
+import org.apache.pulsar.common.policies.data.PublishRate;
+import org.apache.pulsar.common.policies.data.RetentionPolicies;
+import org.awaitility.Awaitility;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+/**
+ * Test KoP message publish throttling.
+ */
+@Slf4j
+public class MessagePublishThrottlingTest extends KopProtocolHandlerTestBase {
+
+    private final boolean preciseTopicPublishRateLimiterEnable;
+
+    public MessagePublishThrottlingTest(boolean preciseTopicPublishRateLimiterEnable) {
+        this.preciseTopicPublishRateLimiterEnable = preciseTopicPublishRateLimiterEnable;
+    }
+
+    public MessagePublishThrottlingTest() {
+        this(false);
+    }
+
+
+    @BeforeClass
+    @Override
+    protected void setup() throws Exception {
+        conf.setPreciseTopicPublishRateLimiterEnable(preciseTopicPublishRateLimiterEnable);
+        super.internalSetup();
+        log.info("success internal setup");
+
+        if (!admin.namespaces().getNamespaces("public").contains("public/__kafka")) {
+            admin.namespaces().createNamespace("public/__kafka");
+            admin.namespaces().setNamespaceReplicationClusters("public/__kafka", Sets.newHashSet("test"));
+            admin.namespaces().setRetention("public/__kafka",
+                    new RetentionPolicies(-1, -1));
+        }
```

Comment:
```suggestion
```

In which case could the namespace `public/__kafka` be not created?

Reply:
Oh, this was copied from another test class. Removed.